nailgun-server/src/main/java/com/martiansoftware/nailgun/NGSessionPool.java

```diff
@@ -63,62 +58,78 @@
 	 * the specified number of threads
 	 * @param server the server to work for
-	 * @param poolsize the maximum number of idle threads to allow
+	 * @param maxIdleSessions the maximum number of idle threads to allow
 	 */
-	NGSessionPool(NGServer server, int poolsize) {
+	NGSessionPool(NGServer server, int maxIdleSessions) {
 		this.server = server;
-		this.poolSize = Math.min(0, poolsize);
-	
-		pool = new NGSession[poolSize];
-		poolEntries = 0;
+		this.maxIdleSessions = Math.max(0, maxIdleSessions);
+
+		idlePool = new LinkedList<>();
+		workingPool = new HashSet<>();
 	}
 
 	/**
 	 * Returns an NGSession from the pool, or creates one if necessary
 	 * @return an NGSession ready to work
 	 */
 	NGSession take() {
-		NGSession result;
 		synchronized(lock) {
-			if (poolEntries == 0) {
-				result = new NGSession(this, server);
-				result.start();
-			} else {
-				--poolEntries;
-				result = pool[poolEntries];
+			if (done) {
+				throw new UnsupportedOperationException("NGSession pool is shutting down");
 			}
+			NGSession session = idlePool.poll();
+			if (session == null) {
+				session = new NGSession(this, server);
+				session.start();
+			}
+			workingPool.add(session);
+			return session;
 		}
-		return (result);
 	}
 	
 	/**
-	 * Returns an NGSession to the pool.  The pool may choose to shutdown
-	 * the thread if the pool is full
+	 * Returns an NGSession to the pool. The pool may choose to shutdown
+	 * the thread if idle pool is full.
 	 * @param session the NGSession to return to the pool
 	 */
 	void give(NGSession session) {
-        boolean shutdown = false;
         synchronized(lock) {
-    		if (done || poolEntries == poolSize) {
-                shutdown = true;
-            } else {
-                pool[poolEntries] = session;
-                ++poolEntries;
-	    	}
+    		if (done) {
+    			// session is already signalled shutdown and removed from all collections
+    			return;
+    		}
+			workingPool.remove(session);
+
+      		if (idlePool.size() < maxIdleSessions) {
+        		idlePool.add(session);
+        		return;
+      		}
         }
-        if (shutdown) session.shutdown();
+       	session.shutdown();
 	}
 	
 	/**
-	 * Shuts down the pool.  Running nails are allowed to finish.
+	 * Shuts down the pool. The function waits for running nails to finish.
 	 */
 	void shutdown() {
+		List<NGSession> allSessions;
 		synchronized(lock) {
 			done = true;
-			while (poolEntries > 0) {
-				take().shutdown();
+			allSessions = Stream.concat(workingPool.stream(), idlePool.stream()).collect(Collectors.toList());
+			idlePool.clear();
+			workingPool.clear();
+		}
+		for (NGSession session : allSessions) {
+			session.shutdown();
+		}
+
+		// wait for all sessions to complete by either returning from waiting state or finishing their nails
+		for (NGSession session : allSessions) {
+			try {
+				session.join();
+			} catch (InterruptedException ex) {
+				// this should never happen so just swallow
```

Comment:
That potentially can hide some problems. What if the server is interrupted? It will try to wait for all the sessions to finish.

I suggest resetting the interrupted flag for the current thread and logging the exception. To me it seems that we need to stop waiting for the rest of the sessions on an interruption.